### ISSUE TITLE
Narrow down the feature sets for resource indexing

### ISSUE THREAD

**Issue by kvark** [gfx-rs/wgpu#1515]
**Is your feature request related to a problem? Please describe.**
We have separate features for uniform buffers, storage buffers, sampled textures, storage textures, each 3 times for different indexing modes. See #1511

**Describe the solution you'd like**
We need to reduce this to a smaller feature set. For example, once feature for buffers, one for textures, one for buffer dynamic indexing, one for texture dynamic indexing.

**Describe alternatives you've considered**
there can be many schemes to split this thing

**Additional context**

**Comment by ElectronicRU** [comment]
Unfortunately, as we've discussed with @cwfitzgerald that is not really feasible: there are platforms which support a subset of those features (ie arrays for sampled but not storage textures), both on Vulkan and on Metal. The mess in the feature flags only reflects the mess in the device support.

**Comment by ElectronicRU** [comment]
Oh and also unless I'm mistaken, the features for storage buffers are currently not there, but yeah it's a matter of time :(

**Comment by cwfitzgerald** [comment]
I suspect we will need to put the DI flags in another location and have have a broader "has a concept of DI" flag here. We may also be able to take a careful study of the ecosystem and group flags into sets that are often shared. We can probably get it down to 3-5 this way.

**Comment by kvark** [comment]
> there are platforms which support a subset of those features (ie arrays for sampled but not storage textures), both on Vulkan and on Metal. The mess in the feature flags only reflects the mess in the device support.

If the fraction of "lost" features is relatively small, we might still want to do this. There is a cost to feature proliferation. It's documentation, testing, maintenance, other cots.

For example, if I search for "DynamicIndexing" features on https://vulkan.gpuinfo.org, it seems rather strongly separated in 2 groups:
  1. uniform buffer + sampled texture (96% android and full desktop)
  2. storage buffer + storage texture (75% on android and full desktop)

For this group, my suggestion would be to leave only 1 feature bit: indexing storage resources. Then make it so the whole "resource array" feature *implies* that uniform resources can be dynamically indexed.

I'm sure we can apply a similar logic to the other features related to resource arrays.

**Comment by ElectronicRU** [comment]
> > there are platforms which support a subset of those features (ie arrays for sampled but not storage textures), both on Vulkan and on Metal. The mess in the feature flags only reflects the mess in the device support.
> 
> If the fraction of "lost" features is relatively small, we might still want to do this. There is a cost to feature proliferation. It's documentation, testing, maintenance, other cots.
> 
> For example, if I search for "DynamicIndexing" features on https://vulkan.gpuinfo.org, it seems rather strongly separated in 2 groups:
> 
> 1. uniform buffer + sampled texture (96% android and full desktop)
> 2. storage buffer + storage texture (75% on android and full desktop)
> 
> For this group, my suggestion would be to leave only 1 feature bit: indexing storage resources. Then make it so the whole "resource array" feature _implies_ that uniform resources can be dynamically indexed.
> 
> I'm sure we can apply a similar logic to the other features related to resource arrays.

Yeah, "arrays of readonly things" and "arrays of writable things" seems to be a much better divide than buffers/textures.
We could add combined flags for now and deprecate the separate ones, then?

**Comment by cwfitzgerald** [comment]
> We could add combined flags for now and deprecate the separate ones, then?

Every wgpu release is breaking, just yeet the old ones.

**Comment by ElectronicRU** [comment]
Oh and also strongly agree with uniting dynamic indexing with arrays - with WGPU workflow, since there's no like specialization constants or header defines, const-indexed arrays are not much different from binding the things separately.

My proposal would be, then:
- uniform buffer and sampled texture arrays ("the OpenGL bindless land", would be a union of those two + dynamic indexing)
- storage buffer and storage texture arrays ("storage resource arrays", again those two + dynamic indexing)
- nonuniform indexing (supported iff for all supported features, nonuniform indexing is also supported).

The only kinky thing I see here is Metal, as there's no support for uniform or storage buffer arrays in MSL, but there is for textures. Could also add a crisscross here, I suppose - add flags for "arrays of buffers" and "arrays of textures" too. Total of 5 flags. (Maybe also add arrays of samplers while we're at it, samplers are also a readonly thing and they fall within the same margin as uniforms and textures iirc)

**Comment by kvark** [comment]
> with WGPU workflow, since there's no like specialization constants or header defines

WebGPU does have specialization constants now, and we need to catch up and implement them

**Comment by kvark** [comment]
I think we can consider this done by #1522
